packages/app-accounts/src/UploadButton.tsx

```diff
@@ -0,0 +1,259 @@
+import ReactFileReader from 'react-file-reader';
+import isUndefined from '@polkadot/util/is/undefined';
+import keyring from '@polkadot/ui-keyring/index';
+
+import AddressMini from '@polkadot/ui-app/AddressMini';
+import Button from '@polkadot/ui-app/Button';
+import Modal from '@polkadot/ui-app/Modal';
+import Unlock from '@polkadot/ui-signer/Unlock';
+
+type State = {
+  address: string,
+  password: string,
+  isPasswordModalOpen: boolean,
+  error?: InputErrorMessage,
+  uploadedFileKeyringPair: KeyringPair$Json | undefined
+};
+
+type Props = I18nProps & BareProps & {
+  icon?: string,
+  isCircular?: boolean,
+  isPrimary?: boolean,
+  size?: Button$Sizes,
+  onChangeAccount: any
+};
+
+class UploadButton extends React.PureComponent<Props, State> {
+  state: State;
+
+  constructor (props: Props) {
+    super(props);
+
+    this.state = this.emptyState();
+  }
+
+  handleUploadedFiles = (files: FileList): void => {
+    const { t } = this.props;
+    const fileList: FileList = files;
+
+    if (!fileList || fileList && !fileList.length) {
+      console.error('Error retrieving file list');
+      return;
+    }
+
+    const fileToUpload: File = fileList[0];
+    const fileReader: FileReader = new FileReader();
+
+    fileReader.onload = (e) => {
+      try {
+        if (!isUndefined(e) && e.target !== null) {
+          const fileContents: any = JSON.parse(e.target.result);
+
+          if (Object.keys(fileContents).includes('address' && 'encoding' && 'meta')) {
+            const json: KeyringPair$Json | undefined = fileContents;
+
+            keyring.addFromJson(json as KeyringPair$Json);
+
+            // Store uploaded wallet in state and open modal to get their password for it
+            this.setState({ uploadedFileKeyringPair: json }, () => this.showPasswordModal());
+          }
+        }
+      } catch (e) {
+        console.error('Error uploading file: ', e);
+        this.setState({ error: { key: t('error'), value: t('Unable to upload account from file') } });
+      }
+    };
+
+    fileReader.readAsText(fileToUpload);
+  }
+
+  processUploadedFileStorage = (): void => {
+    const { password, uploadedFileKeyringPair } = this.state;
+    const { t, onChangeAccount } = this.props;
+
+    if (isUndefined(uploadedFileKeyringPair) || !uploadedFileKeyringPair.address) {
+      return;
+    }
+
+    const json = uploadedFileKeyringPair;
+
+    // Reset password so it is not pre-populated on the form on subsequent uploads
+    this.setState({ password: '' }, () => {
+      try {
+        if (!json || !Object.keys(json).length) {
+          return;
+        }
+
+        const pairRestored = keyring.restoreAccount(json, password);
+
+        if (pairRestored) {
+          this.hidePasswordModal();
+          onChangeAccount(pairRestored.publicKey());
+
+          return;
+        } else {
+          this.setState({ error: { key: t('error'), value: t('Unable to upload account into memory') } });
```

Comment:
setState inside setState callback, this can't be right.

Seem to be making _exactly_ the same comments in 2 files here - DRY it up and share as needed, this seems like a lot of copy-and-pasting going on.